### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by sipa**
Pick and choose.

### CODE REVIEW

File: src/primitives/block.h [hunk 1]
```diff
@@ -77,6 +77,8 @@ class CBlock : public CBlockHeader
     std::vector<CTransaction> vtx;
 
     // memory only
+    // TODO: cache check status inside the validation logic instead of inside of
+    // the primitive data type CBlock.
```
Comment (TheBlueMatt) [inline src/primitives/block.h]: This misses half the point of the comment - that we probably /do/ want to have the ability to cache the things that CBlock could check for itself (like the merkle tree).

File: src/blockencodings.h [hunk 2]
```diff
@@ -22,7 +22,8 @@ struct TransactionCompressor {
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(tx); //TODO: Compress tx encoding
+        // Future transaction compression standards could be implemented here
```
Comment (TheBlueMatt) [inline src/blockencodings.h]: I'd still call it actionable, you just have to bump the cmpctblock protocol version and write a BIP to do so.

**Comment by sipa** [comment]
CBlock should not check anything, it is a data structure. Validation logic does not belong in data structures.

File: src/blockencodings.h (hunk 2 above)
Comment (sipa) [inline src/blockencodings.h]: It's not something someone looking at that code can do, it needs a proposal
first. So yes, it's a TODO but not at the code level.

**Comment by TheBlueMatt** [comment]
Difference of opinion :). I find Bitcoin Core's general design of "everything is a dumb data structure and all the actual logic goes in main.cpp so its all in one place" to be annoying :p

File: src/blockencodings.cpp [hunk 3]
```diff
@@ -101,9 +101,6 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
     const std::vector<std::pair<uint256, CTxMemPool::txiter>>& vTxHashes = pool->vTxHashes;
     for (size_t i = 0; i < vTxHashes.size(); i++) {
         uint64_t shortid = cmpctblock.GetShortID(vTxHashes[i].first);
-        //TODO: Put an ifdef guard around the prefetch here
-        if (i + 1 < vTxHashes.size())
```
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: This change seemed to increase performance by a small margin, please do not suggest removing without benchmarks.

File: src/blockencodings.cpp [hunk 4]
```diff
@@ -30,12 +30,7 @@ CBlockHeaderAndShortTxIDs::CBlockHeaderAndShortTxIDs(const CBlock& block) :
 }
 
 void CBlockHeaderAndShortTxIDs::FillShortTxIDSelector() const {
-    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
-    stream << header << nonce;
-    CSHA256 hasher;
-    hasher.Write((unsigned char*)&(*stream.begin()), stream.end() - stream.begin());
-    uint256 shorttxidhash;
-    hasher.Finalize(shorttxidhash.begin());
+    uint256 shorttxidhash = (CSHA256Writer(SER_NETWORK, PROTOCOL_VERSION) << header << nonce).GetHash();
```
Comment (TheBlueMatt) [inline src/blockencodings.cpp]: I'm gonna opt to leave this change here for now...I agree with it in principal, but would prefer to limit code churn as much as possible at this point. Can you submit it as a separate pull req later?

File: src/main.cpp [hunk 5]
```diff
@@ -4541,7 +4541,7 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                         assert(!"cannot load block from disk");
                     if (inv.type == MSG_CMPCT_BLOCK)
                     {
-                        if (mi->second->nHeight >= chainActive.Height() - 10) {
+                        if (mi->second->nHeight >= chainActive.Height() - MAX_BLOCKS_TO_ANNOUNCE) {
```
Comment (TheBlueMatt) [inline src/main.cpp]: MAX_BLOCKS_TO_ANNOUNCE (confusing name...) has a very different meaning...I'd prefer to either leave 10 or introduce a new constant.

Comment (TheBlueMatt) [inline src/main.cpp]: For now I'll add a comment.

**Comment by sipa** [comment]
Well validation logic obviously has to be split up and modularized further
and not remain all in main.cpp.

File: src/blockencodings.cpp (hunk 3 above)
Comment (sipa) [inline src/blockencodings.cpp]: Thanks, if it's the result of a benchmark I'll believe you. I had assumed
that this was a leftover from when it required reading from various places
in memory.

File: src/blockencodings.cpp (hunk 4 above)
Comment (sipa) [inline src/blockencodings.cpp]: Ok.

File: src/main.cpp (hunk 5 above)
Comment (sipa) [inline src/main.cpp]: Right. I wanted to (and think we still should) make it announce using
cmpctblocks even when there are multiple blocks to announce at once (so we
don't lose BIP152's advantages completely in small reorgs), but it turned
out to be harder than I thought.

**Comment by sipa** [comment]
In any case, a TODO is not the place to suggest a significant design
change. That's confusing to people reading it who may want to pick it up.

**Pull request closed without merging**